machine/v1/types_controlplanemachineset.go

```diff
@@ -0,0 +1,312 @@
+	// +optional
+	Placement AWSFailureDomainPlacement `json:"placement,omitempty"`
+}
+
+// AWSFailureDomainPlacement configures the placement information for the AWSFailureDomain.
+type AWSFailureDomainPlacement struct {
+	// AvailabilityZone is the availability zone of the instance.
+	// +kubebuilder:validation:Required
+	AvailabilityZone string `json:"availabilityZone"`
+}
+
+// AzureFailureDomain configures failure domain information for the Azure platform.
+type AzureFailureDomain struct {
+	// Availability Zone for the virtual machine.
+	// If nil, the virtual machine should be deployed to no zone.
+	// +kubebuilder:validation:Required
+	Zone string `json:"zone"`
+}
+
+// GCPFailureDomain configures failure domain information for the GCP platform
+type GCPFailureDomain struct {
+	// Zone is the zone in which the GCP machine provider will create the VM.
+	// +kubebuilder:validation:Required
+	Zone string `json:"zone"`
+}
+
+// OpenStackFailureDomain configures failure domain information for the OpenStack platform
+type OpenStackFailureDomain struct {
+	// The availability zone from which to launch the server.
+	// +kubebuilder:validation:Required
+	AvailabilityZone string `json:"availabilityZone"`
+}
+
+// ControlPlaneMachineSetStatus represents the status of the ControlPlaneMachineSet CRD.
+type ControlPlaneMachineSetStatus struct {
+	// Conditions represents the observations of the ControlPlaneMachineSet's current state.
+	// Known .status.conditions.type are: Available, Degraded and Progressing.
+	// +patchMergeKey=type
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// ObservedGeneration is the most recent generation observed for this
+	// ControlPlaneMachineSet. It corresponds to the ControlPlaneMachineSets's generation,
+	// which is updated on mutation by the API Server.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Replicas is the number of Control Plane Machines created by the
+	// ControlPlaneMachineSet controller.
+	// Note that during update operations this value may differ from the
+	// desired replica count.
+	// +optional
+	Replicas int32 `json:"replicas,omitempty"`
+
+	// ReadyReplicas is the number of Control Plane Machines created by the
+	// ControlPlaneMachineSet controller which are ready.
+	// +optional
+	ReadyReplicas int32 `json:"readyReplicas,omitempty"`
+
+	// UpdatedReplicas is the number of non-terminated Control Plane Machines
+	// created by the ControlPlaneMachineSet controller that have the desired
+	// provider spec.
+	// +optional
+	UpdatedReplicas int32 `json:"updatedReplicas,omitempty"`
+
+	// UpdatedReadyReplicas is the number of non-terminated Control Plane Machines
+	// created by the ControlPlaneMachineSet controller that have the desired
+	// provider spec and are ready.
+	// +optional
+	UpdatedReadyReplicas int32 `json:"updatedReadyReplicas,omitempty"`
```

Comment:
I spoke with @JoelSpeed about it on slack, the feedback I gave him is that this field seems unnecessary, since comparing controlplanemachineset with deployments and how it drives its rollout it should be possible to set `UpdatedReplicas=0` when a new rollout starts, and update `ReadyReplicas` whenever the new machine is fully operational. There is no need to differentiate between `UpdatedReadyReplicas` and `ReadyReplicas`. It's important to note that during rollout `ReadyReplicas` can have values bigger than `.spec.Replicas` which is caused by the current surge of 1, and that should be call out explicitly in the field's description.

Reply:
The suggestion of adding this field was originally from @damdo. By adding it, we make it very explicit to users the number of machines with the current desired spec and of those, how many are ready. However, as this design is based heavily on deployments, I can see the arguments for keeping it more similar to the deployment API.

We ran through a bunch of examples on a call and in each scenario I believe you can infer from the combination of Ready, Updated and Replicas how many replicas are updated or not. So we should be fine without this for now.

It does create some confusion within the codebase because it's hard to define updated as "updated and ready" in the API, when the rest of the code it means something else, but we can deal with that

Reply:
+1.
Thanks for your input on this @soltysh.